src/charm.py

```diff
@@ -38,6 +36,7 @@
 from ops.model import ActiveStatus, BlockedStatus, Container, ModelError, WaitingStatus
 from ops.pebble import ExecError, Layer
 
+from charm_state import CharmConfigInvalidError, CharmState
```

Comment:
I thought we'll be initializing `CharmState` (or `CharmConfig` as mentioned in another comment) using `from_charm` in the constructor, to get the validated config and then we'll get values from there instead of `return self.model.config.get("cni-type")`. 
Is there any reason why we're not doing this?

Reply:
It seems that using `self._charm_config = CharmConfig.from_charm(charm=self)` the the charm init won't work: at that point the `charm.config` reflects the values before updating the config. One solution could be to instantiate a `CharmConfig` instance on `config-changed` and `pebble-ready`. Wdyt?

Reply:
I'm not sure what you mean by `reflects the values before updating the config`? Every time an event happens, new instance of the charm is created. It seems to me, that when the constructor is called, it'll always read the current config of the charm, because it's the config change that triggers the event not the other way around.
Please correct me if I'm missing something.

Reply:
As discussed in these days, having the `self._charm_config` in the constructor may not be enough, as in some cases (`ops.Harness` during testing and deferred events in production) the charm is not reinstantiated. For this reason, I override the object in the `config-changed` and `pebble-ready` events for now.